flink-clients/src/main/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrap.java

```diff
@@ -165,7 +170,9 @@ public void stop() {
                                 if (applicationStatus == ApplicationStatus.CANCELED
                                         || applicationStatus == ApplicationStatus.FAILED) {
                                     LOG.info("Application {}: ", applicationStatus, t);
-                                    return dispatcherGateway.shutDownCluster(applicationStatus);
+                                    return shouldShutDownOnFinish
+                                            ? dispatcherGateway.shutDownCluster(applicationStatus)
+                                            : CompletableFuture.completedFuture(Acknowledge.get());
                                 }
                             }
 
```

Comment:
Ah, good point. I didn't change the logic in this branch because no shutdown was triggered in this class before the change. I suppose that shutdown is triggered in some other way if the future completes exceptionally (which is still the case since `thenCompose` does not handle the exception, as you observed)

I tend to think that we want to keep the cluster alive even in case of unexpected failures. Do you think this could cause problems?

Reply:
The shutdown in that case is usually triggered by the errorHandler, which by default shuts down the cluster (org.apache.flink.runtime.entrypoint.ClusterEntrypoint#onFatalError).

I think that this is also fine and, after thinking about it some more, we should keep that.
If the main() method throws an exception we never actually reach a stage where a `job` exists in the JM. Users wouldn't be able to query the state anyway; .at most they could check the logs and then decide whether to restart or not, but logs aren't really a stable interface so building anything on top of that seems pointless.
So we just treat it like any error in the startup procedure, and just fail the process.

Reply:
I added a comment stating that we do not prevent shutdown in that case